### ISSUE TITLE
Comparison Operator 

### ISSUE THREAD

**Issue by lawrencenull** [lucaong/minisearch#119]
This is a fantastic library, I was wondering if there is currently a way to filter using numeric comparison operators? I need to add filters for number ranges, pageviews etc.

<html>
<body>

Comparison Operator | Description
-- | --
>= | Greater Than or Equal
< | Less Than
<= | Less Than or Equal
IN ( ) | Matches a value in a list

</body>
</html>

**Comment by lucaong** [comment]
Hi @lawrencenull ,
while MiniSearch only indexes fields as strings of text, one way to achieve what you need is to use the `filter` option: you can supply a function that receives each search result, and can return true or false to keep it or drop it from the final results. It is usually convenient to add the numeric field in the `storeFields` option, so it’s immediately available in the results to be filtered.

```javascript
const miniSearch = new MiniSearch({
  fields: ["title", "artist"],
  storeFields: ["year"]
})

// Say documents are like:
// { title: "Love Burns", artist: "BRMC", year: 2001 }
miniSearch.addAll(documents)

miniSearch.search("love", {
  filter: (result) => result.year > 2000
})
```

This way is the most common, and recommended if the numeric filter is applied on top of a text search, so it filters out an already reduced number of results. You can see it in action in the [demo code](https://github.com/lucaong/minisearch/blob/232965e195c8d6af73b396e61ffae65ef70e9b50/examples/src/app.js#L109) when filtering for start/end year.

Note that if you _only_ filter by numeric range, and do not perform any full-text search, there are probably better tools than MiniSearch.

**Comment by lawrencenull** [comment]
Thank you for your reply, I am doing a full text search but was wanting to filter and display by popularity. This should work perfect.

**Comment by lawrencenull** [comment]
I found a very small library that works perfect for my use case, allowing sorting objects by multiple fields. I set up a quick demo here http://jsfiddle.net/lawrencenull/74adzvLc/5/ in case anyone else needs to do something similar. This allows sorting after filtering and can be useful for data drill-downs or rendering a front end UI. Minified it is 699 characters, so a very light addition.

**Comment by 684efs3** [comment]
> numeric



> Hi @lawrencenull , thanks for your kind words!
> 
> While MiniSearch only indexes fields as strings of text, one way to achieve what you need is to use the `filter` option: you can supply a function that receives each search result, and can return true or false to keep it or drop it from the final results. It is usually convenient to add the numeric field in the `storeFields` option, so it’s immediately available in the results to be filtered.
> 
> ```js
> const miniSearch = new MiniSearch({
>   fields: ["title", "artist"],
>   storeFields: ["year"]
> })
> 
> // Say documents are like:
> // { title: "Love Burns", artist: "BRMC", year: 2001 }
> miniSearch.addAll(documents)
> 
> miniSearch.search("love", {
>   filter: (result) => result.year > 2000
> })
> ```
> 
> This way is the most common, and recommended if the numeric filter is applied on top of a text search, so it filters out an already reduced number of results. You can see it in action in the [demo code](https://github.com/lucaong/minisearch/blob/232965e195c8d6af73b396e61ffae65ef70e9b50/examples/src/app.js#L109) when filtering for start/end year.
> 
> Note that if you _only_ filter by numeric range, and do not perform any full-text search, there are probably better tools than MiniSearch.

Is there a way to do this if sometimes there is a search query and sometimes there isn't? In the case where there is no search query I would like to only perform the numerical filter. Is this possible?

**Comment by lucaong** [comment]
Hi @684efs3 ,
at the moment, the best way to implement what you want is to filter your documents "manually" if the search is empty, something like:

```javascript
const miniSearch = new MiniSearch({
  fields: ["title", "artist"],
  storeFields: ["year"]
})

// Say documents are like:
// { title: "Love Burns", artist: "BRMC", year: 2001 }
miniSearch.addAll(documents)

// Map documents by their ID:
const docById = documents.reduce((byId, doc) => {
  byId[doc.id] = doc
  return byId
})

let filteredDocuments = null

// Somewhere in your app, check if there is a non-empty search query, and act accordingly:
if (query && query.length > 0) {
  // There is a search query - get search results and filter them
  const searchResults = miniSearch.search(query, {
    filter: (result) => result.year > 2000
  })
  filteredDocuments = searchResults.map((result) => docById[result.id])
} else {
  // No search query - simply filter the whole document list
  filteredDocuments = documents.filter((doc) => doc.year > 2000)
}
```

That said, this gives me an idea for a possible upcoming feature of `MiniSearch`, enabling users to get all documents (with no search query), possibly applying filtering, boosting, etc. While this is feasible in user code, it might still be convenient to do that within `MiniSearch`. In case I will implement such feature, I will update this thread.

**Comment by lucaong** [comment]
@684efs3 this pull request adds the feature you are requesting: https://github.com/lucaong/minisearch/pull/234

Basically, in your case, when there is no search query you can pass `MiniSearch.wildcard` as the first argument to `miniSearch.search()` to match all documents, but still apply filtering and document boosting.

The example above becomes:

```javascript
const miniSearch = new MiniSearch({
  fields: ["title", "artist"],
  storeFields: ["year"]
})

// Say documents are like:
// { title: "Love Burns", artist: "BRMC", year: 2001 }
miniSearch.addAll(documents)

// Somewhere in your app, if the query is missing or blank, use the wildcard instead
if (query == null || query.length === 0) {
  query = MiniSearch.wildcard
}

const searchResults = miniSearch.search(query, {
  filter: (result) => result.year > 2000
})
```

The pull request is not merged yet, but I will likely merge it soon and release a new version.

**Comment by lucaong** [comment]
Released on NPM as part of `v6.2.0`.

**Comment by 684efs3** [comment]
Thank you very much!!! Have a great weekend!
